drv/cosmo-hf/src/apob.rs

```diff
@@ -122,3 +129,660 @@ impl From<HfError> for ApobError {
+        meta: Meta,
+        data: Option<ApobRawPersistentData>,
+    },
+    WrotePersistentData {
+        meta: Meta,
+        data: ApobRawPersistentData,
+    },
+    HashMismatch {
+        expected_hash: [u8; 32],
+        actual_hash: [u8; 32],
+    },
+    ApobSlotErase {
+        slot: ApobSlot,
+    },
+    ApobSlotEraseDone {
+        slot: ApobSlot,
+        time_ms: u64,
+    },
+    ApobSlotSectorErase {
+        slot: ApobSlot,
+        offset: u32,
+    },
+    BadApobSig {
+        expected: [u8; 4],
+        actual: [u8; 4],
+    },
+    BadApobVersion {
+        expected: u32,
+        actual: u32,
+    },
+    BadApobSize {
+        expected: u32,
+        actual: u32,
+    },
+    BadApobWalk {
+        expected: u32,
+        actual: u32,
+    },
+}
+ringbuf!(Trace, 16, Trace::None);
+
+#[derive(Copy, Clone, PartialEq)]
+pub(crate) enum ApobSlot {
+    Slot0,
+    Slot1,
+}
+
+impl core::ops::Not for ApobSlot {
+    type Output = Self;
+    fn not(self) -> Self::Output {
+        match self {
+            ApobSlot::Slot0 => ApobSlot::Slot1,
+            ApobSlot::Slot1 => ApobSlot::Slot0,
+        }
+    }
+}
+
+impl ApobSlot {
+    pub fn base_addr(&self) -> FlashAddr {
+        match self {
+            ApobSlot::Slot0 => FlashAddr::new(APOB_SLOT0_ADDR).unwrap(),
+            ApobSlot::Slot1 => FlashAddr::new(APOB_SLOT1_ADDR).unwrap(),
+        }
+    }
+
+    pub fn flash_addr(&self, offset: u32) -> Option<FlashAddr> {
+        let base = self.base_addr();
+        if offset >= APOB_SLOT_SIZE {
+            return None;
+        }
+        base.0.checked_add(offset).and_then(FlashAddr::new)
+    }
+}
+
+#[derive(Copy, Clone, PartialEq)]
+pub(crate) enum ApobState {
+    /// Waiting for `ApobStart`
+    Waiting {
+        write_slot: ApobSlot,
+        read_slot: Option<ApobSlot>,
+    },
+    /// Receiving and writing data to host flash
+    Ready {
+        write_slot: ApobSlot,
+        expected_length: u32,
+        expected_hash: ApobHash,
+        any_written: bool,
+    },
+    /// Writing data to flash is no longer allowed
+    Locked {
+        /// We store the first commit result for idempotency, because the host
+        /// is allowed to retry the `ApobCommit` message.  Subsequent commits
+        /// return the same result.
+        commit_result: Result<(), ApobCommitError>,
+    },
+}
+
+/// Persistent data, stored in Bonus Flash to select an APOB slot
+#[derive(
+    Copy, Clone, Eq, PartialEq, IntoBytes, FromBytes, Immutable, KnownLayout,
+)]
+#[repr(C)]
+pub struct ApobRawPersistentData {
+    /// Must always be `APOB_PERSISTENT_DATA_MAGIC`.
+    oxide_magic: u32,
+
+    /// Must always be `APOB_PERSISTENT_DATA_HEADER_VERSION` (for now)
+    header_version: u32,
+
+    /// Monotonically increasing counter
+    pub monotonic_counter: u64,
+
+    /// Either 0 or 1; directly translatable to [`ApobSlot`]
+    pub slot_select: u32,
+
+    /// CRC-32 over the rest of the data using the iSCSI polynomial
```

Comment:
Any significance to the choice of polynomial? (curiosity more than anything)

Reply:
I copied what we used for [`HfRawPersistentData`](https://github.com/oxidecomputer/hubris/blob/2dc17d22155df3e999bc4ba24534ea69881ad3f1/drv/hf-api/src/lib.rs#L141-L159) 😄